packages/just-task/src/cache.ts

```diff
@@ -0,0 +1,72 @@
+import { getPackageDeps } from '@microsoft/package-deps-hash';
+import { argv } from './option';
+import { resolveCwd } from './resolve';
+import fs from 'fs-extra';
+import path from 'path';
+import { logger } from 'just-task-logger';
+
+const cachedTask: string[] = [];
+
+export function registerCachedTask(taskName: string) {
+  cachedTask.push(taskName);
+}
+
+export function isCached(taskName: string) {
+  if (cachedTask.indexOf(taskName) < 0) {
+    return false;
+  }
+
+  const currentHash = getHash(taskName);
+  const cachePath = getCachePath();
+  const cacheFile = path.join(cachePath, 'package-deps.json');
+
+  if (!fs.existsSync(cacheFile)) {
+    return false;
+  }
+
+  if (!fs.pathExistsSync(cachePath)) {
+    fs.mkdirpSync(cachePath);
+  }
+
+  try {
+    const cachedHash = JSON.parse(fs.readFileSync(path.join(cachePath, 'package-deps.json')).toString());
+
+    // TODO: make a more robust comparison
+    return JSON.stringify(currentHash) === JSON.stringify(cachedHash);
+  } catch (e) {
+    logger.warn('Invalid package-deps.json detected');
+  }
+
+  return false;
+}
+
+export function saveCache(taskName: string) {
+  if (cachedTask.indexOf(taskName) < 0) {
+    return;
+  }
+
+  const cachePath = getCachePath();
+
+  if (!fs.pathExistsSync(cachePath)) {
+    fs.mkdirpSync(cachePath);
+  }
+
+  fs.writeFileSync(path.join(cachePath, 'package-deps.json'), JSON.stringify(getHash(taskName)));
+}
+
+function getCachePath() {
+  const packageJsonFilePath = resolveCwd('package.json')!;
+  const rootPath = path.dirname(packageJsonFilePath);
+  return path.join(rootPath, 'node_modules/.just');
+}
+
+function getHash(taskName: string) {
+  const { $0: scriptNameArg, ...args } = argv();
+  const hash = getPackageDeps();
```

Comment:
what does this do for mono repos with respect to shrinkwraps/lock files? For example updating to the latest react in the lockfile/shrinkwrap should invalidate my cache, even if there are no code changes (e.g. if the package json already used semver like "react: ^16.8.0")

Reply:
Dealing with mono repos directly isn't really the job of just really... I think what I need is a mechanism for a monorepo task to invalidate cache when lock files are deleted (clearing cache isn't something I have in this PR... so I wouldn't say this is the final API to be integrated yet)

Reply:
Okay I added a getPackageDeps to use the .git root path to capture changes everywhere. Then I filter the files to just the ones UNDER the current cwd (the package dir). And then I add back in the yarn.lock, pnpmfile.js, shrinkwrap.yml and package-lock.json